File: languages/ruby/lib/oso/polar/query.rb
# frozen_string_literal: true

require 'json'

module Oso
  module Polar
    # A single Polar query.
    class Query # rubocop:disable Metrics/ClassLength
      include Enumerable

      # @param ffi_query [FFI::Query]
      # @param host [Oso::Polar::Host]
      def initialize(ffi_query, host:, bindings: {})
        @calls = {}
        @ffi_query = ffi_query
        ffi_query.enrich_message = host.method(:enrich_message)
        @host = host
        bindings.each { |k, v| ffi_query.bind k, host.to_polar(v) }
      end

      # Create an enumerator that can be polled to advance the query loop. Yields
      # results one by one.
      #
      # @yieldparam [Hash<String, Object>]
      # @return [Enumerator]
      # @raise [Error] if any of the FFI calls raise one.
      def each(&block)
        run(&block)
      end

      private

      # @return [Hash<Integer, Enumerator>]
      attr_reader :calls
      # @return [FFI::Query]
      attr_reader :ffi_query
      # @return [Host]
      attr_reader :host

      # Send result of predicate check across FFI boundary.
      #
      # @param result [Boolean]
      # @param call_id [Integer]
      # @raise [Error] if the FFI call raises one.
      def question_result(result, call_id:)
        ffi_query.question_result(result, call_id: call_id)
      end

      # Send next result of Ruby method call across FFI boundary.
      #
      # @param result [String]
      # @param call_id [Integer]
      # @raise [Error] if the FFI call raises one.
      def call_result(result, call_id:)
        ffi_query.call_result(result, call_id: call_id)
      end

      # Retrieve the next result from a registered call and pass it to {#to_polar}.
      #
      # @param id [Integer]
      # @return [Hash]
      # @raise [StopIteration] if the call has been exhausted.
      def next_call_result(id)
        host.to_polar(calls[id].next)
      end

      # Send application error across FFI boundary.
      #
      # @param message [String]
      # @raise [Error] if the FFI call raises one.
      def application_error(message)
        ffi_query.application_error(message)
      end

      # Fetch the next result from calling a Ruby method and prepare it for
      # transmission across the FFI boundary.
      #
      # @param attribute [#to_sym]
      # @param call_id [Integer]
      # @param instance [Hash<String, Object>]
      # @param args [Array<Hash>]
      # @param kwargs [Hash<String, Object>]
      # @raise [Error] if the FFI call raises one.
      def handle_call(attribute, call_id:, instance:, args:, kwargs:) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
        instance = host.to_ruby(instance)
        rel = get_relationship(instance.class, attribute)
        return handle_relationship(call_id, instance, rel) unless rel.nil?

        args = args.map { |a| host.to_ruby(a) }
        kwargs = Hash[kwargs.map { |k, v| [k.to_sym, host.to_ruby(v)] }]
        # The kwargs.empty? check is for Ruby < 2.7.
        result = if kwargs.empty?
                   instance.__send__(attribute, *args)
                 else
                   instance.__send__(attribute, *args, **kwargs)
                 end
        call_result(host.to_polar(result), call_id: call_id)
      rescue ArgumentError, NoMethodError => e
        application_error(e.message)
        call_result(nil, call_id: call_id)
      end

      # Get the type information for a field on a class.
      #
      # @param cls [UserType]
      # @param tag [String]
      # @return [UserType]
      # @raise [Error] if no information is found
      def get_field(cls, tag) # rubocop:disable Metrics/AbcSize
        raise unless cls.fields.key? tag

        ref = cls.fields[tag]
        return host.types[ref] unless ref.is_a? ::Oso::Polar::DataFiltering::Relation

        case ref.kind
        when 'one'
          host.types[ref.other_type]
        when 'many'
          host.types[Array]
        end
      end

      # Check if a series of dot operations on a base class yield an
      # instance of another class.
      def handle_external_isa_with_path(data) # rubocop:disable Metrics/AbcSize
        sup = host.types[data['class_tag']]
        bas = host.types[data['base_tag']]
        path = data['path'].map(&host.method(:to_ruby))
        sub = path.reduce(bas) { |cls, tag| get_field(cls, tag) }
        answer = sub.klass.get <= sup.klass.get
        question_result(answer, call_id: data['call_id'])
      rescue StandardError => e
        application_error e.message
        question_result(nil, call_id: data['call_id'])
      end

      def handle_next_external(call_id, iterable)
        unless calls.key? call_id
          value = host.to_ruby iterable
          raise InvalidIteratorError unless value.is_a? Enumerable

          calls[call_id] = value.lazy
        end

        result = next_call_result(call_id)
        call_result(result, call_id: call_id)
      rescue StopIteration
        call_result(nil, call_id: call_id)
      end

      def handle_make_external(data) # rubocop:disable Metrics/AbcSize
        id = data['instance_id']
        raise DuplicateInstanceRegistrationError, id if host.instance? id

        constructor = data['constructor']['value']
        raise InvalidConstructorError unless constructor.key? 'Call'

        cls_name = constructor['Call']['name']
        args = constructor['Call']['args'].map { |arg| host.to_ruby(arg) }
        kwargs = constructor['Call']['kwargs'] || {}
        kwargs = Hash[kwargs.map { |k, v| [k.to_sym, host.to_ruby(v)] }]
        host.make_instance(cls_name, args: args, kwargs: kwargs, id: id)
      end

      # Run the main Polar loop, yielding results as they are emitted from the VM.
      #
      # @yieldparam [Hash<String, Object>]
      # @return [Enumerator]
      # @raise [Error] if any of the FFI calls raise one.
      def run # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
        loop do # rubocop:disable Metrics/BlockLength
          event = ffi_query.next_event
          case event.kind
          when 'Done'
            break
          when 'Result'
            yield event.data['bindings'].transform_values { |v| host.to_ruby(v) }
          when 'MakeExternal'
            handle_make_external(event.data)
          when 'ExternalIsaWithPath'
            handle_external_isa_with_path(event.data)
          when 'ExternalCall'
            call_id = event.data['call_id']
            instance = event.data['instance']
            attribute = event.data['attribute']
            args = event.data['args'] || []
            kwargs = event.data['kwargs'] || {}
            handle_call(attribute, call_id: call_id, instance: instance, args: args, kwargs: kwargs)
          when 'ExternalIsSubSpecializer'
            instance_id = event.data['instance_id']
            left_tag = event.data['left_class_tag']
            right_tag = event.data['right_class_tag']
            answer = host.subspecializer?(instance_id, left_tag: left_tag, right_tag: right_tag)
            question_result(answer, call_id: event.data['call_id'])
          when 'ExternalIsSubclass'
            call_id = event.data['call_id']
            left = event.data['left_class_tag']
            right = event.data['right_class_tag']
            answer = host.subclass?(left_tag: left, right_tag: right)
            question_result(answer, call_id: call_id)
          when 'ExternalIsa'
            instance = event.data['instance']
            class_tag = event.data['class_tag']
            answer = host.isa?(instance, class_tag: class_tag)
            question_result(answer, call_id: event.data['call_id'])
          when 'Debug'
            msg = event.data['message']
            if msg
              msg = host.enrich_message(msg) if msg
              puts msg
            end
            print 'debug> '
            begin
              input = $stdin.readline.chomp.chomp(';')
            rescue EOFError
              next
            end
            command = JSON.dump(host.to_polar(input))
            ffi_query.debug_command(command)
          when 'ExternalOp'
            op = event.data['operator']
            args = event.data['args'].map(&host.method(:to_ruby))
            answer = host.operator(op, args)
            question_result(answer, call_id: event.data['call_id'])
          when 'NextExternal'
            call_id = event.data['call_id']
            iterable = event.data['iterable']
            handle_next_external(call_id, iterable)
          else
            raise "Unhandled event: #{JSON.dump(event.inspect)}"
          end
        end
      end

      def get_relationship(cls, attr)
        typ = host.types[cls]
        return unless typ

        rel = typ.fields[attr]
        return unless rel.is_a? ::Oso::Polar::DataFiltering::Relation

        rel
      end

      def handle_relationship(call_id, instance, rel) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
        typ = host.types[rel.other_type]

        cls = typ.klass.get

        condition = ::Oso::Polar::Data::Filter::Condition.new(
          ::Oso::Polar::Data::Filter::Projection.new(cls, rel.other_field),
          'Eq',
          instance.send(rel.my_field)
        )

        filter = ::Oso::Polar::Data::Filter.new(
          model: cls,
          relations: [],
          conditions: [[condition]],
          types: host.types
        )

        res = host.adapter.execute_query host.adapter.build_query(filter)

        if rel.kind == 'one'
          raise "multiple parents: #{res}" if res.length > 1

          res = res[0] if res.length == 1
        end

        call_result(host.to_polar(res), call_id: call_id)
      end
    end
  end
end

File: languages/python/oso/polar/query.py
from collections.abc import Iterable
from typing import List, Any, Generator, Dict
import json

from .exceptions import (
    InvalidIteratorError,
    InvalidCallError,
    InvalidConstructorError,
    PolarRuntimeError,
)
from .data_filtering import Relation
from .data import DataFilter, Condition, Projection

NATIVE_TYPES = [int, float, bool, str, dict, type(None), list]


class QueryResult:
    """Response type of a call to the `query` API"""

    def __init__(self, results: List[Any]) -> None:
        self.success = len(results) > 0
        self.results = [r["bindings"] for r in results]
        self.traces = [r["trace"] for r in results]


class Query:
    """Execute a Polar query through the FFI/event interface."""

    def __init__(self, ffi_query, *, host=None, bindings=None):
        self.ffi_query = ffi_query
        self.ffi_query.set_message_enricher(host.enrich_message)
        self.host = host
        self.calls = {}
        for (k, v) in (bindings or {}).items():
            self.bind(k, v)

    def __del__(self) -> None:
        del self.host
        del self.ffi_query

    def bind(self, name, value):
        """Bind `name` to `value` for the duration of the query."""
        self.ffi_query.bind(name, self.host.to_polar(value))

    def run(self) -> Generator[Dict[str, Any], None, None]:
        """Run the event loop and yield results."""
        assert self.ffi_query, "no query to run"
        while True:
            ffi_event = self.ffi_query.next_event()
            event = json.loads(ffi_event)
            kind = [*event][0]
            data = event[kind]

            call_map = {
                "MakeExternal": self.handle_make_external,
                "ExternalCall": self.handle_external_call,
                "ExternalOp": self.handle_external_op,
                "ExternalIsa": self.handle_external_isa,
                "ExternalIsaWithPath": self.handle_external_isa_with_path,
                "ExternalIsSubSpecializer": self.handle_external_is_subspecializer,
                "ExternalIsSubclass": self.handle_external_is_subclass,
                "NextExternal": self.handle_next_external,
                "Debug": self.handle_debug,
            }

            if kind == "Done":
                break
            elif kind == "Result":
                bindings = {
                    k: self.host.to_python(v) for k, v in data["bindings"].items()
                }
                trace = data["trace"]
                yield {"bindings": bindings, "trace": trace}
            elif kind in call_map:
                call_map[kind](data)
            else:
                raise PolarRuntimeError(f"Unhandled event: {json.dumps(event)}")

    def handle_make_external(self, data):
        instance_id = data["instance_id"]
        constructor = data["constructor"]["value"]
        if "Call" not in constructor:
            raise InvalidConstructorError()
        cls_name = constructor["Call"]["name"]
        args = [self.host.to_python(arg) for arg in constructor["Call"]["args"]]
        kwargs = constructor["Call"]["kwargs"] or {}
        kwargs = {k: self.host.to_python(v) for k, v in kwargs.items()}
        self.host.make_instance(cls_name, args, kwargs, instance_id)

    def handle_relation(self, instance, rel):

        other_cls = self.host.types[rel.other_type].cls
        condition = Condition(
            Projection(other_cls, rel.other_field),
            "Eq",
            getattr(instance, rel.my_field),
        )
        data_filter = DataFilter(other_cls, [], [[condition]], self.host.types)
        adapter = self.host.adapter
        query = adapter.build_query(data_filter)
        results = adapter.execute_query(query)

        if rel.kind == "one":
            assert len(results) == 1
            return results[0]
        elif rel.kind == "many":
            return results
        else:
            raise ValueError(rel.kind)

    def handle_external_call(self, data):
        call_id = data["call_id"]
        instance = self.host.to_python(data["instance"])

        attribute = data["attribute"]

        # Lookup the attribute on the instance.
        try:
            # Check if it's a relationship
            attr = None
            cls = instance.__class__
            if cls in self.host.types:
                cls_rec = self.host.types[cls]
                typ = cls_rec.fields
                if attribute in typ:
                    attr_typ = typ[attribute]
                    if isinstance(attr_typ, Relation):
                        attr = self.handle_relation(instance, attr_typ)

            if attr is None:
                attr = getattr(instance, attribute)
        except AttributeError as e:
            self.ffi_query.application_error(str(e))
            self.ffi_query.call_result(call_id, None)
            return
        if (
            callable(attr) and data["args"] is not None
        ):  # If it's a function, call it with the args.
            args = [self.host.to_python(arg) for arg in data["args"]]
            kwargs = data["kwargs"] or {}
            kwargs = {k: self.host.to_python(v) for k, v in kwargs.items()}
            result = attr(*args, **kwargs)
        elif data["args"] is not None:
            raise InvalidCallError(
                f"tried to call '{attribute}' but it is not callable"
            )
        else:  # If it's just an attribute, it's the result.
            result = attr

        # Return the result of the call.
        self.ffi_query.call_result(call_id, self.host.to_polar(result))

    def handle_external_op(self, data):
        op = data["operator"]
        args = [self.host.to_python(arg) for arg in data["args"]]
        answer = self.host.operator(op, args)
        self.ffi_query.question_result(data["call_id"], answer)

    def handle_external_isa(self, data):
        instance = data["instance"]
        class_tag = data["class_tag"]
        answer = self.host.isa(instance, class_tag)
        self.ffi_query.question_result(data["call_id"], answer)

    def handle_external_isa_with_path(self, data):
        base_tag = data["base_tag"]
        path = data["path"]
        class_tag = data["class_tag"]
        try:
            answer = self.host.isa_with_path(base_tag, path, class_tag)
            self.ffi_query.question_result(data["call_id"], answer)
        except AttributeError as e:
            # TODO(gj): make sure we are printing but not failing on receipt of
            # this error in core.
            self.ffi_query.application_error(str(e))
            self.ffi_query.question_result(data["call_id"], False)

    def handle_external_is_subspecializer(self, data):
        instance_id = data["instance_id"]
        left_tag = data["left_class_tag"]
        right_tag = data["right_class_tag"]
        answer = self.host.is_subspecializer(instance_id, left_tag, right_tag)
        self.ffi_query.question_result(data["call_id"], answer)

    def handle_external_is_subclass(self, data):
        left_tag = data["left_class_tag"]
        right_tag = data["right_class_tag"]
        answer = self.host.is_subclass(left_tag, right_tag)
        self.ffi_query.question_result(data["call_id"], answer)

    def handle_next_external(self, data):
        call_id = data["call_id"]
        iterable = data["iterable"]

        if call_id not in self.calls:
            value = self.host.to_python(iterable)
            if isinstance(value, Iterable):
                self.calls[call_id] = iter(value)
            else:
                raise InvalidIteratorError(f"{value} is not iterable")

        # Return the next result of the call.
        try:
            value = next(self.calls[call_id])
            self.ffi_query.call_result(call_id, self.host.to_polar(value))
        except StopIteration:
            self.ffi_query.call_result(call_id, None)

    def handle_debug(self, data):
        if data["message"]:
            print(self.host.enrich_message(data["message"]))
        try:
            command = input("debug> ").strip(";")
        except EOFError:
            command = "continue"
        self.ffi_query.debug_command(self.host.to_polar(command))
